routing: Add triert package with Xor Trie backed routing table

iand opened the pull request:
Fixes #44 

Trie-backed routing table, safe for concurrent use. Refresh capabilities will come with #45

Notes for follow up:

 - the routing table has no enforced maximum size
 - it may be populated with a crawl mechanism like FullRT but this process will be decoupled from the routing table
 - ~it might be better to copy the xor trie into the go-kademlia package to allow extension, specifically storing both the key and the node id in the trie for efficiency. Also we can then use our own Key type rather than relying on the coincidence that both are `[]byte`~

iand commented on routing/triert/table.go:
```diff
@@ -0,0 +1,228 @@
+		if err != nil {
+			return 0, err
+		}
+		if keyCpl == cpl {
```
Should be  >=

guillaumemichel commented on routing/triert/table.go:
```diff
@@ -0,0 +1,228 @@
+	"sync"
+	"sync/atomic"
+
+	// "github.com/libp2p/go-libp2p-xor/kademlia"
```
```suggestion
```

guillaumemichel commented on routing/triert/table.go:
```diff
@@ -0,0 +1,228 @@
+func (rt *TrieRT) AddPeer(ctx context.Context, node address.NodeID) (bool, error) {
+	kk := node.Key()
+	if kk.Size() != rt.self.Size() {
+		return false, ErrKeyWrongLength
```
`ErrKeyWrongLength` could be defined in the package `routing` as it is a generic error for operations on routing tables.

guillaumemichel replied:
Sorry, I missed
```go
// TODO: move to key package and replace ErrInvalidKey
// ErrKeyWrongLength indicates a key has the wrong length
var ErrKeyWrongLength = errors.New("key has wrong length")
```
moving the error to the `key` package also makes sense.

iand replied:
Yes, I didn't want to do this without discussion. I think we should replace `ErrInvalidKey` by `ErrKeyWrongLength` for two reasons:

1. a key of a different length isn't an invalid key, it's just the wrong length so the error is clearer
2. ErrInvalidKey isn't a single error so it's currently hard to use with `errors.Is`. We could improve that with a custom error type but I think it's simpler to have a specific error to signal incompatible lengths
3.

guillaumemichel replied:
I agree that it would make more sense to replace `ErrInvalidKey` with something like `ErrWrongKeyLength`.

I don't have much experience with defining error types in Go, all suggestions are welcome 😄  In error testing, I mostly used `require.Equal(t, ExpectedError, err)`, but I am not sure that it is the best practice

iand replied:
Currently I'm using `ErrMismatchedKeyLength` from the `key/trie` package

guillaumemichel commented on routing/triert/table.go:
```diff
@@ -0,0 +1,227 @@
+}
+
+func closestAtDepth(ctx context.Context, k tkey.Key, t *trie.Trie, depth int, n int, found []tkey.Key) ([]tkey.Key, error) {
+	if ctx.Err() != nil {
```
is `ctx` ever expected to be cancelled during an operation?

iand replied:
Generally I like to allow cancellation on long loops to improve responsiveness, but since this is doing very little work in each iteration then I agree it's not needed.

guillaumemichel commented on routing/triert/table.go:
```diff
@@ -0,0 +1,227 @@
+	self key.KadKey
+
+	// keymu is held during keys and keyNodes mutations to serialize changes
+	keymu sync.Mutex
```
Technically, if this routing table implementation is used by a single thread scheduler, it wouldn't need a lock at all.

Having a lock makes it thread safe, and usable in multi threads scheduler contexts, but we don't have such schedulers yet.

iand replied:
I assume there will be other clients of the routing table that will need to do reads. For example the Kubo CLI expects to be able to dump all the entries in a routing table

guillaumemichel replied:
OK, in this case Kubo CLI wouldn't need to take the DHT's lock for it, and would need to enqueue the routing table read action to the scheduler 👍🏻

iand replied:
Removed concurrency support from the TrieRT

guillaumemichel commented on routing/triert/table.go:
```diff
@@ -0,0 +1,227 @@
+}
+
+// AddPeer tries to add a peer to the routing table
+func (rt *TrieRT) AddPeer(ctx context.Context, node address.NodeID) (bool, error) {
```
tracing is usually welcome

e.g
```go
	ctx, span := util.StartSpan(ctx, "AddPeer",
		trace.WithAttributes(attribute.Stringer("NodeID", node)),
	)
	defer span.End()
```

iand replied:
yes, will add

guillaumemichel commented on routing/triert/table.go:
```diff
@@ -0,0 +1,227 @@
+
+	return nodes, nil
+}
+
+func closestAtDepth(ctx context.Context, k tkey.Key, t *trie.Trie, depth int, n int, found []tkey.Key) ([]tkey.Key, error) {
+	if ctx.Err() != nil {
+		return found, ctx.Err()
+	}
+
+	// If we've already found enough peers, abort.
+	if n == len(found) {
+		return found, nil
+	}
+
+	// Find the closest direction.
+	dir := k.BitAt(depth)
+	var chosenDir byte
+	if t.Branch[dir] != nil {
+		// There are peers in the "closer" direction.
+		chosenDir = dir
+	} else if t.Branch[1-dir] != nil {
+		// There are peers in the "less closer" direction.
+		chosenDir = 1 - dir
+	} else if t.Key != nil {
+		// We've found a leaf
+		return append(found, t.Key), nil
+	} else {
+		// We've found an empty node?
+		return found, nil
+	}
+
+	var err error
+	// Add peers from the closest direction first, then from the other direction.
+	found, err = closestAtDepth(ctx, k, t.Branch[chosenDir], depth+1, n, found)
+	if err != nil {
+		return found, err
+	}
+
+	return closestAtDepth(ctx, k, t.Branch[1-chosenDir], depth+1, n, found)
+}
```
Simplifying `NearestPeers` 
```suggestion
// NearestPeers returns the n closest peers to a given key
func (rt *TrieRT) NearestPeers(ctx context.Context, kk key.KadKey, n int) ([]address.NodeID, error) {
	if kk.Size() != rt.self.Size() {
		return nil, ErrKeyWrongLength
	}

	keys := rt.keys.Load().(*trie.Trie)
	closestKeys := closestAtDepth(tkey.Key(kk), keys, 0, n)
	keyNodes := rt.keyNodes.Load().(map[string]address.NodeID)

	nodes := make([]address.NodeID, 0, len(closestKeys))
	for _, c := range closestKeys {
		if id, ok := keyNodes[string(key.KadKey(c))]; ok {
			nodes = append(nodes, id)
		}
	}

	return nodes, nil
}

func closestAtDepth(k tkey.Key, t *trie.Trie, depth int, n int) []tkey.Key {
	// Find the closest direction.
	dir := k.BitAt(depth)
	if t.Key != nil {
		// We've found a leaf
		return []tkey.Key{t.Key}
	} else if t.Branch[0] == nil && t.Branch[1] == nil {
		// We've found an empty node?
		return nil
	}

	// Add peers from the closest direction first
	found := closestAtDepth(k, t.Branch[dir], depth+1, n)
	if len(found) == n {
		return found
	}

	// Didn't find enough peers in the closest direction, try the other direction.
	return append(found, closestAtDepth(k, t.Branch[1-dir], depth+1, n-len(found))...)
}
```

iand replied:
Nice!

guillaumemichel commented on routing/triert/table_test.go:
```diff
@@ -0,0 +1,375 @@
+	id  string
+	key key.KadKey
+}
+
```
```suggestion
var _ address.NodeID = (*node)(nil)

```

guillaumemichel commented on routing/triert/table.go:
```diff
@@ -0,0 +1,227 @@
+}
+
+// CplSize returns the number of peers in the table that share the specified common prefix length with the table's key.
+func (rt *TrieRT) CplSize(cpl int) int {
```
This functions return the number of peers with a CPL >= `cpl`. It should return the number of peers with a CPL == `cpl`.

While it is true that `0000` and `0001` share a prefix of length `1`, of length `2` and of length `3`, we define the CPL as the longest common prefix. Hence the only CPL of `0000` and `0001` is `3`.

Assuming there is only `0001` in `0000`'s routing table.
`"0000".CplSize(1) -> 0` and `"0000".CplSize(3) -> 1`

iand replied:
Fixed

dennis-tra commented on routing/triert/table.go:
```diff
@@ -0,0 +1,227 @@
+}
+
+// CplSize returns the number of peers in the table that share the specified common prefix length with the table's key.
+func (rt *TrieRT) CplSize(cpl int) int {
```
Judging from the comments, this should do the same: https://github.com/libp2p/go-libp2p-xor/blob/1a5ca6fd64c49fc92364b47a005b1734d53bf5f8/kademlia/bucket.go#L10 and looks a bit more concise.

However, judging from the code, it seems that both are doing different things :D 

> // BucketAtDepth returns the bucket in the routing table at a given depth.
> // A bucket at depth D holds contacts that share a prefix of exactly D bits with node.

vs

> // CplSize returns the number of peers in the table that share the specified common prefix length with the table's key.

iand replied:
Rewritten this function but I agree it's similar. But since we're moving the xor trie into the repo we won't be using the old xor/kademlia package

guillaumemichel commented on routing/triert/table_test.go:
```diff
@@ -0,0 +1,375 @@
+	return key.KadKey(buf)
+}
+
+type node struct {
```
`github.com/plprobelab/go-kademlia/network/address/kadid` could be used here. It was built for testing routing behaviors.

iand replied:
I looked at it but it doesn't allow me to control the key and the string value independently. I wanted the strings to be readable for debugging tests.

guillaumemichel commented on routing/triert/table.go:
```diff
@@ -0,0 +1,227 @@
+	return closestAtDepth(ctx, k, t.Branch[1-chosenDir], depth+1, n, found)
+}
+
+func (rt *TrieRT) Find(ctx context.Context, kk key.KadKey) (address.NodeID, error) {
```
It could be used to verify if a peer is already in the routing table or not, e.g in the context of https://github.com/libp2p/go-libp2p-kad-dht/issues/811 where we send a kademlia request to the remote peer before adding it to the routing table.

In go-libp2p-kbucket this logic is implemented in https://github.com/libp2p/go-libp2p-kbucket/blob/6d85d4e63aa3b8e04b9b55e1de64be8a52532f21/table.go#L119. So there is no use case now (except testing), but it may be useful in the future.

iand replied:
It's used in tests and potentially in other diagnostics / consistency checking. It's not part of the Table interface.

dennis-tra replied:
If it's possible we could un-export it.

iand replied:
I don't see the need to unexport it. It's a useful function for code that want to directly query the table

guillaumemichel commented on routing/triert/table.go:
```diff
@@ -0,0 +1,227 @@
+// RemoveKey tries to remove a peer identified by its Kademlia key from the
+// routing table
+func (rt *TrieRT) RemoveKey(ctx context.Context, kk key.KadKey) (bool, error) {
+	if kk.Size() != rt.self.Size() {
```
Users should not operate with different KadKey lengths. IMO it is good to check it to avoid bad surprises later.

iand replied:
It's a pity that Go doesn't allow generic types to be specified using specific slice lengths

dennis-tra replied:
I also hoped we could constrain the usage to one specific key type.

guillaumemichel commented on routing/triert/table.go:
```diff
@@ -0,0 +1,227 @@
+	keyNodes atomic.Value
+}
+
+func New(self key.KadKey) *TrieRT {
```
This routing table is unlimited in size. It would be great to limit the size of each bucket, if we go for the _normal server_ routing table implementation.

iand replied:
I think it would be useful to have full control over the maximum population of each bucket. We could actually implement this as a filter. Allow a function to be specified with the following signature that is called by the rt in AddPeer:

```Go
func(rt *TrieRT, kk key.KadKey) bool
```

Then we can pass a filter that limits buckets to 20
```Go
func Bucket20(rt *TrieRT, kk key.KadKey) bool {
    cpl := rt.Cpl(kk)  // this doesn't exist but it basically does rt.self.CommonPrefixLength(kk)
    return rt.CplSize(cpl) < 20
}
```

Or one that allows unlimited peers: 

```Go
func Unlimited(rt *TrieRT, kk key.KadKey) bool {
    return true
}
```

Or one that varies occupancy by cpl:

```Go
func VaryOccupancy(rt *TrieRT, kk key.KadKey) bool {
    cpl := rt.Cpl(kk)
    return rt.CplSize(cpl) < 5*cpl
}
```

iand replied:
Done

iand commented on routing/triert/table.go:
```diff
@@ -0,0 +1,227 @@
+
+	// keys holds a pointer to an immutable trie
+	// any store to keys must be performed while holding keymu, loads may be performed without the lock
+	keys atomic.Value
```
It allows lock free reads. Only the mutation path needs to take a lock. My assumption here is that the number of reads of a routing table is far higher than the number of mutations.

dennis-tra replied:
I actually doubt that this will be the case 🤔 we're trying to add peers to the routing table after every successful `FIND_NODE`/`GET_PROVIDERS` RPC (see `query.go:L196`) and try to remove them after every failed `FIND_NODE`/`GET_PROVIDERS` RPC (see `query.go:L250`).

So we have many writes vs a single read per query. Unless we have other places where we frequently read from the routing table I believe that the assumption doesn't hold. We could batch the writes towards the end of a query which makes the #write ~= #reads.

iand replied:
The current implementation makes batching writes very simple since all the mutation work is done while holding a single lock and then replacing the immutable table.

As an aside. I don't think the query should be directly mutating the routing table. It should schedule the fact that a new peer has been found to be responsive/not responsive. I could argue that this then becomes the responsibility of the refresh mechanism to probe the peer and determine whether it belongs in the routing table.

Currently the routing table is simply a map of known peer ids sorted by distance, without any liveness information. I would combine this with a refresh mechanism,  a diversity filter, a bootstrap mechanism and possibly a crawler. I don't think those functions belong in the table itself.

It's possible that we can arrange things so that we don't need concurrency control since the event queue and scheduler perform that role.

dennis-tra replied:
> The current implementation makes batching writes very simple since all the mutation work is done while holding a single lock and then replacing the immutable table.

This would require a new API, right? Because the current API locks and releases for each peer addition.

We could also give `AddPeer` a variadic type parameter.

> I don't think the query should be directly mutating the routing table.

💯! This is also what I've done, in my event loop implementation.

iand replied:
> This would require a new API, right? Because the current API locks and releases for each peer addition.

Yes. AddPeers or make it variadic.

iand replied:
Removed concurrency support from the TrieRT

iand commented on routing/triert/table.go:
```diff
@@ -0,0 +1,227 @@
+	return nodes, nil
+}
+
+func closestAtDepth(ctx context.Context, k tkey.Key, t *trie.Trie, depth int, n int, found []tkey.Key) ([]tkey.Key, error) {
```
this is because we're using methods from the xor tree which defines its own key type

iand replied:
we're going to copy your generic trie into the repo which will eliminate this problem

iand commented:
The imported trie with extra data field meant I could drop the map between key and node in the triert

iand commented:
Error checking is in preparation for https://github.com/plprobelab/go-kademlia/pull/51#issuecomment-1621703741 if we decide to take that route

dennis-tra commented on key/keyutil/keyutil.go:
```diff
@@ -1,19 +1,19 @@
 	"github.com/plprobelab/go-kademlia/key"
 )
 
+var rng = rand.New(rand.NewSource(299792458))
```
speed of light? :D

iand commented:
PTAL

Recent changes:

 - added config to TrieRT
 - added KeyFilterFunc
 - removed concurrency handling from TrieRT, it mutates the Trie in place
 - added benchmarks

Initial benchmarks:

```
goos: linux
goarch: amd64
pkg: github.com/plprobelab/go-kademlia/key/trie
cpu: 13th Gen Intel(R) Core(TM) i7-13650HX
BenchmarkBuildTrieMutable/1000-20          5368      239713 ns/op    1286779 ns/key      186240 B/op        2910 allocs/op
BenchmarkBuildTrieMutable/10000-20          278     4296195 ns/op     119433 ns/key     1836428 B/op       28694 allocs/op
BenchmarkBuildTrieMutable/100000-20          39    29098613 ns/op      11348 ns/key    18480013 B/op      288750 allocs/op
BenchmarkBuildTrieImmutable/1000-20        2350      692483 ns/op    1627331 ns/key      787776 B/op       12309 allocs/op
BenchmarkBuildTrieImmutable/10000-20         96    13987724 ns/op     134282 ns/key     9982505 B/op      155976 allocs/op
BenchmarkBuildTrieImmutable/100000-20        10   103134639 ns/op      10313 ns/key    121270315 B/op     1894848 allocs/op
BenchmarkAddMutable/1000-20               25603       47707 ns/op    4885774 ns/key       46465 B/op         726 allocs/op
BenchmarkAddMutable/10000-20                806     1463271 ns/op     471757 ns/key      466178 B/op        7284 allocs/op
BenchmarkAddMutable/100000-20               100    10893145 ns/op      43572 ns/key     4603648 B/op       71932 allocs/op
BenchmarkAddImmutable/1000-20              9826      232817 ns/op    9150635 ns/key      232640 B/op        3635 allocs/op
BenchmarkAddImmutable/10000-20              297     3863193 ns/op     458946 ns/key     2744902 B/op       42889 allocs/op
BenchmarkAddImmutable/100000-20              52    23498035 ns/op      48876 ns/key    32888064 B/op      513876 allocs/op
BenchmarkRemoveMutable/1000-20            37214       34692 ns/op    5164169 ns/key           0 B/op           0 allocs/op
BenchmarkRemoveMutable/10000-20             960     1309283 ns/op     502763 ns/key           4 B/op           0 allocs/op
BenchmarkRemoveMutable/100000-20            100    10223293 ns/op      40893 ns/key           1 B/op           0 allocs/op
BenchmarkRemoveImmutable/1000-20           8344      249415 ns/op    8324454 ns/key      194496 B/op        3039 allocs/op
BenchmarkRemoveImmutable/10000-20           284     4433920 ns/op     503692 ns/key     2476353 B/op       38693 allocs/op
BenchmarkRemoveImmutable/100000-20           42    28681932 ns/op      48186 ns/key    29994752 B/op      468668 allocs/op
BenchmarkFindPositive/1000-20          18115773       63.32 ns/op                             0 B/op           0 allocs/op
BenchmarkFindPositive/10000-20         11876832       95.95 ns/op                             0 B/op           0 allocs/op
BenchmarkFindPositive/100000-20         5064666       232.6 ns/op                             0 B/op           0 allocs/op
BenchmarkFindNegative/1000-20          35327665       34.01 ns/op                             0 B/op           0 allocs/op
BenchmarkFindNegative/10000-20         15040240       75.56 ns/op                             0 B/op           0 allocs/op
BenchmarkFindNegative/100000-20         6080850       198.8 ns/op                             0 B/op           0 allocs/op


goos: linux
goarch: amd64
pkg: github.com/plprobelab/go-kademlia/routing/triert
cpu: 13th Gen Intel(R) Core(TM) i7-13650HX
BenchmarkBuildTable/1000-20              6938    272436 ns/op   1890159 ns/node    185072 B/op     2892 allocs/op
BenchmarkBuildTable/10000-20              256   4579850 ns/op   117244 ns/node    1836924 B/op    28702 allocs/op
BenchmarkBuildTable/100000-20              40   9681654 ns/op   11873 ns/node    18486147 B/op   288846 allocs/op
BenchmarkFindPositive/1000-20        20827332     55.88 ns/op                           0 B/op        0 allocs/op
BenchmarkFindPositive/10000-20       14346126     83.10 ns/op                           0 B/op        0 allocs/op
BenchmarkFindPositive/100000-20       6442490     189.8 ns/op                           0 B/op        0 allocs/op
BenchmarkFindNegative/1000-20        35187136     33.13 ns/op                           0 B/op        0 allocs/op
BenchmarkFindNegative/10000-20       15310812     73.62 ns/op                           0 B/op        0 allocs/op
BenchmarkFindNegative/100000-20       6700594     181.9 ns/op                           0 B/op        0 allocs/op
BenchmarkNearestPeers/1000-20          318604      5927 ns/op                        5965 B/op       44 allocs/op
BenchmarkNearestPeers/10000-20         129309      9016 ns/op                        5970 B/op       44 allocs/op
BenchmarkNearestPeers/100000-20        229111      5404 ns/op                        5983 B/op       44 allocs/op
BenchmarkChurn/1000-20                5453337     199.6 ns/op                          64 B/op        1 allocs/op
BenchmarkChurn/10000-20               3280557     375.0 ns/op                          63 B/op        0 allocs/op
BenchmarkChurn/100000-20              3017536     388.5 ns/op                          65 B/op        1 allocs/op

```

iand commented:
Windows/macos tests are still flaky. I think this is quic related.

dennis-tra commented on key/key_test.go:
```diff
@@ -111,3 +111,17 @@ func TestCompare(t *testing.T) {
+	require.Equal(t, 1, kk.BitAt(6))
+	require.Equal(t, 1, kk.BitAt(7))
+	require.Equal(t, 1, kk.BitAt(8))
+	require.Equal(t, 0, kk.BitAt(15))
```
not sure if it makes sense to also assert the behavior for invalid input like `i < 0` and `i >= len(bits(kk))`. Just something that crossed my mind.

iand replied:
will do

dennis-tra commented on key/keyutil/keyutil.go:
```diff
@@ -0,0 +1,51 @@
+package keyutil
```
suggestion: could move to the `key` package into a `rand.go` file.

If it's just used for testing we could also move it to the `testutil` package.

iand replied:
I think I will move to testutil. I don't think we want to encourage creating random keys with math/rand for anything other than tests

dennis-tra commented on key/keyutil/keyutil.go:
```diff
@@ -0,0 +1,51 @@
+}
+
+// RandomWithPrefix returns a KadKey of length l having a prefix equal to the bit pattern held in s.
+func RandomWithPrefix(s string, l int) key.KadKey {
```
- the comment should mention the maximum prefix of 64 bits.
- since there are input checks there could also be a check that `len(s) <= l*8` is true

iand replied:
done

dennis-tra commented on key/keyutil/keyutil.go:
```diff
@@ -0,0 +1,51 @@
+
+var rng = rand.New(rand.NewSource(299792458))
+
+// Random returns a KadKey of length l populated with random data.
```
here, `length` refers to the number of bytes instead of the number of bits. I haven't checked there rest but the usage of `length` in the context of a KadKey should be consistent throughout the code.

Alternatively we could use:

- `size` -> bytes
- `length` -> bits

iand replied:
good catch

dennis-tra commented on key/trie/trie.go:
```diff
@@ -0,0 +1,323 @@
+	"github.com/plprobelab/go-kademlia/key"
+)
+
+var ErrMismatchedKeyLength = errors.New("key length does not match existing keys")
```
rm in favor of the `key` package error

iand replied:
I think it belongs here. The key package error is for invalid keys. Here all the keys are valid keys, but one does not match the format of the others

dennis-tra commented on key/trie/trie.go:
```diff
@@ -0,0 +1,323 @@
+	return &Trie[T]{}
+}
+
+func (tr *Trie[T]) Key() key.KadKey {
```
Not sure about the implications but `Trie` conforms to the `NodeID` interface. Perhaps rather export the `Key` and `Data` fields?

iand replied:
I unexprted Key and Data when adding the mutable Add/Remove methods. My thinking is that we want to limit the chances for users to corrupt the trie by overwriting fields like `key`. In practice the Key method will be inlined so there's no real performance overhead.

Not sure I understand the point about conforming to NodeID interface - what would be the use case?

dennis-tra commented on key/trie/trie.go:
```diff
@@ -0,0 +1,323 @@
+
+// Add adds the key to trie, returning a new trie.
+// Add is immutable/non-destructive: the original trie remains unchanged.
+func Add[T any](tr *Trie[T], kk key.KadKey, data T) (*Trie[T], error) {
```
Do we need both, the mutable and immutable versions of the API?

dennis-tra commented on routing/triert/table.go:
```diff
@@ -0,0 +1,180 @@
+	return n
+}
+
+func countCpl(t *nodeTrie, kk key.KadKey, cpl int, depth int) (int, error) {
```
This could move to trie.Trie

dennis-tra commented on routing/triert/table.go:
```diff
@@ -0,0 +1,180 @@
+	data address.NodeID
+}
+
+func closestAtDepth(kk key.KadKey, t *nodeTrie, depth int, n int) []entry {
+	if t.IsLeaf() {
```
This could move to `trie.Trie` and return the trie instead of an `entry`
